### PR TITLE
fix(model): prevent post_mp_layer.dims mutation across k-fold runs

### LINKED ISSUE #169: Bug: post_mp_layer.dims mutated across k-fold runs causing corrupted MLflow params
## Summary

The `post_mp_layer.dims` config is mutated in-place during model construction, causing each k-fold iteration to append an additional `dim_out=1` to the same list. This corrupts both the model architecture and the logged MLflow parameters.

## Root Cause

In `stgym/model.py`, lines 26 and 51:

```python
cfg.post_mp_layer.dims.append(dim_out)
```

This mutates the shared config object. When running k-fold CV, the same config is reused across folds, so:
- Fold 0: `[32, 16]` → appends 1 → `[32, 16, 1]`
- Fold 1: `[32, 16, 1]` → appends 1 → `[32, 16, 1, 1]`
- Fold 2: `[32, 16, 1, 1]` → appends 1 → `[32, 16, 1, 1, 1]`
- ...

## Evidence

From MLflow analysis of `postmp` experiment:

| Expected design choices | Actual logged values |
|------------------------|----------------------|
| `['16', '32', '64', '64,32', '32,16']` | 25 unique values including `[32, 16, 1, 1, 1, 1]` |

Within the same `group_id`, we see progressively corrupted values:
```
group_id 18:
[32, 16]             → 0 trailing 1s
[32, 16, 1]          → 1 trailing 1
[32, 16, 1, 1]       → 2 trailing 1s
[32, 16, 1, 1, 1]    → 3 trailing 1s
[32, 16, 1, 1, 1, 1] → 4 trailing 1s
```

## Impact

1. **Model architecture corruption** — Later folds have extra MLP layers
2. **MLflow params corrupted** — Cannot aggregate results by design choice since each fold has different logged value
3. **RCT analysis invalid** — Rank comparisons across design choices are broken

## Fix

Replace mutation with immutable operation:

```python
# Before (BAD)
cfg.post_mp_layer.dims.append(dim_out)

# After (GOOD)
dims = cfg.post_mp_layer.dims + [dim_out]
```

Then use `dims` instead of `cfg.post_mp_layer.dims` when constructing the MLP.

## Testing Requirements

Add unit tests to verify:
1. Config object is not mutated after model construction
2. Multiple model instantiations with same config produce identical architectures
3. K-fold simulation confirms dims remain unchanged across folds

## Files to Modify

- `stgym/model.py` (lines 26, 51)
- Add test in `tests/test_model.py` or `tests/test_config_immutability.py`

### PR DESCRIPTION
**Pull request by xiaohan2012**
## Problem

`cfg.post_mp_layer.dims.append(dim_out)` in `STGraphClassifier` and `STNodeClassifier` mutates the shared `PostMPConfig` object. In k-fold CV, the same config is reused across folds, causing dims to accumulate trailing `dim_out` values each fold — corrupting both model architecture and logged MLflow params.

Fixes #169

## Solution

Replace in-place mutation with `model_copy(update=...)` to create a per-instantiation copy of the config with `dim_out` appended. The original config is untouched.

## Changes

- `stgym/model.py` — replace `cfg.post_mp_layer.dims.append(dim_out)` with `post_mp_config = cfg.post_mp_layer.model_copy(update={"dims": cfg.post_mp_layer.dims + [dim_out]})` in both `STGraphClassifier` and `STNodeClassifier`
- `tests/test_model.py` — add `TestConfigImmutability` parametrized over both model classes: verifies dims unchanged after single construction and after 5 repeated constructions (k-fold simulation)

## Testing

- [x] `pre-commit run --all-files` passes
- [x] `pytest tests/ -m 'not slow'` passes (235 tests)
- [x] `TestConfigImmutability` confirms dims stay unchanged after single and repeated (×5) model construction

Co-Authored-By: Claude Sonnet 4.6 <noreply@anthropic.com>

### CODE REVIEW

**Review by xiaohan2012 — COMMENTED** [review]
## Review: fix(model): prevent post_mp_layer.dims mutation across k-fold runs

**Verdict:** APPROVE
**Issue:** #169

### Blocking

None

### Warnings

> **[warning]** `tests/test_model.py:46-83` (Code Quality)
> `TestConfigImmutability` is a parametrized class but its parametrize decorator lists anonymous lambdas, making test IDs like `STGraphClassifier-<lambda>` hard to parse in CI failure output.
> **Suggestion:** Add `ids=["graph_clf", "node_clf"]` to the `@pytest.mark.parametrize` call so test IDs read `test_dims_unchanged_after_construction[graph_clf]` instead of `test_dims_unchanged_after_construction[STGraphClassifier-<lambda>]`.

> **[warning]** `tests/test_model.py:69-83` (Test Coverage)
> The tests only verify that `cfg.post_mp_layer.dims` is unchanged; they do not assert that the constructed model's MLP actually receives the correct final dims (i.e., `[32, num_classes]`). Without that assertion the test is purely a mutation guard, not a correctness check of the output dimension.
> **Suggestion:** After construction, assert that the model's final linear layer has `out_features == self.num_classes` (or equivalent) to verify the model was wired correctly with the appended `dim_out`.

### Nits

> **[nit]** `stgym/model.py:26-28, 53-55` (Code Quality)
> The two `model_copy(update={"dims": ...})` blocks in `STGraphClassifier.__init__` and `STNodeClassifier.__init__` are identical. A small private helper would eliminate the duplication.
> **Suggestion:**
> ```python
> def _append_dim_out(post_mp_cfg: PostMPConfig, dim_out: int) -> PostMPConfig:
>     return post_mp_cfg.model_copy(update={"dims": post_mp_cfg.dims + [dim_out]})
> ```

### Test Coverage

The two new test methods in `TestConfigImmutability` are well-structured and directly reproduce the k-fold failure scenario (repeated construction with the same config). Parametrization correctly covers both model classes. Coverage gap noted above: the model's internal `post_mp` layer output dimension is not verified, so a latent off-by-one in how `dim_out` feeds into the copy would not be caught. The existing forward-pass tests in `TestSTGraphClassifier` and `TestSTNodeClassifier` partially compensate by checking output tensor shapes.

### Summary

The fix is correct: replacing in-place `list.append` on a shared Pydantic model with `model_copy(update=...)` is the idiomatic Pydantic v2 approach, and I verified manually that the new list is fully independent from the original. All 235 existing tests pass, the 4 new tests pass, and ruff reports no linting or formatting issues. The two warnings are minor improvements to test readability and coverage that are not blockers. This PR is safe to merge.

**Comment by xiaohan2012** [comment]
Addressed reviewer feedback:
- Added `ids=["graph_clf", "node_clf"]` to parametrize for readable test IDs
- Replaced parameter-order-based shape check with a forward pass asserting `pred.shape[-1] == num_classes` — more robust and tests actual end-to-end dim wiring (note: inner linear uses `torch_geometric.nn.dense.linear.Linear`, not `torch.nn.Linear`, so module-traversal approach didn't work)
